File: 5회차/programmers_algorithm/3.py
def solution(number, k):
    answer = ''
    n = len(number)
    size = n-k
    pnt = 0
    dp = [[0 for i in range(10)] for j in range(n+1)]
    for i in range(1, n+1):
        for j in range(0, 10):
            dp[i][j]=dp[i-1][j]
        dp[i][int(number[i-1])]+=1
        
    pnt = 0
    while pnt < n and len(answer)<size:
        l = pnt
        r = n-(size-len(answer)-1)
        for i in range(9, -1, -1):
            if dp[r][i]-dp[l][i]>0:
                answer=answer+str(i)
                break
        while number[pnt]!=str(i):
            pnt+=1
        pnt+=1
    return answer

File: 9회차/answer/3.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int dx[4]={0,1,0,-1};
int dy[4]={1,0,-1,0};

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    int a[rows][columns];
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            a[i][j] = i*columns+j+1;
        }
    }
    vector<int> answer;
    
    
    for (int i = 0; i < queries.size(); i++) {
        int x1,y1,x2,y2;
        x1=queries[i][0]; y1=queries[i][1]; x2=queries[i][2]; y2=queries[i][3];
        x1--; y1--; x2--; y2--;
        int ans = 2000000;
        int pnt = 0;
        
        int x = x1; int y = y1;
        int rem = a[x][y];
        while (true) {
            //rem = a[x][y];
            x = x+dx[pnt];
            y = y+dy[pnt];
            if (x<x1 || x>x2 || y<y1 || y>y2) {
                if (pnt==3) break;
                x=x-dx[pnt];
                y=y-dy[pnt];
                pnt++;
                x=x+dx[pnt];
                y=y+dy[pnt];
            }
            int trem = a[x][y];
            a[x][y] = rem;
            rem = trem;
            ans = min(ans,rem);
        }
        answer.push_back(ans);
    }
    
    return answer;
}